File: ocaml/11-20/18.ml
(*Extract a slice from a list. (medium)

Given two indices, i and k, the slice is the list containing the elements between the i'th and k'th element of the original list (both limits included). Start counting the elements with 0 (this is the way the List module numbers elements).
*)

let rec take n xs = match (n, xs) with
  | (_, []) | (0, _) -> []
  | (_, y::ys) -> y::(take (n-1) ys)

let rec skip n xs = match (n, xs) with
  | (_, []) | (0, _) -> xs
  | (_, y::ys) -> skip (n-1) ys

let slice xs from upto = skip from xs |> take (upto - from + 1)


let _ = assert ((slice ["a";"b";"c";"d";"e";"f";"g";"h";"i";"j"] 2 6) = ["c"; "d"; "e"; "f"; "g"])

File: prolog/11-20/18.pl
/*
Extract a slice from a list.
    Given two indices, I and K, the slice is the list containing the elements between the I'th and K'th element of the original list (both limits included). Start counting the elements with 1.

    Example:
    ?- slice([a,b,c,d,e,f,g,h,i,k],3,7,L).
     L = [c,d,e,f,g]

*/

take(0, _, Acc, Ys) :- reverse(Acc, Ys).
take(N, [H|T], Acc, Ys) :- 
  N1 is N - 1,
  take(N1, T, [H|Acc], Ys).
  
drop(0, Xs, Xs).
drop(N, [_|T], Ys) :-
  N > 0,
  N1 is N - 1,
  drop(N1, T, Ys).

slice([], _, _, []).
slice(Xs, I, K, Zs) :- 
  K >= I,
  I1 is I - 1,
  KI is K - I,
  drop(I1, Xs, Ys),
  take(KI, Ys, [], Zs).

